MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java

```diff
@@ -166,60 +161,89 @@ public CancellableTask getProducts(final List<Long> productIds, final Callback<L
-        }
-        if (sortOrder == null) {
-            throw new NullPointerException("sortOrder cannot be null");
-        }
 
         return retrofitService
-            .getProducts(appId, collectionId, pageSize, page, sortOrder.toString())
+            .getProductTagPage(appId, page, pageSize)
             .retryWhen(networkRetryPolicyProvider.provide())
             .doOnNext(new RetrofitSuccessHttpStatusCodeHandler<>())
-            .compose(new UnwrapRetrofitBodyTransformer<ProductListings, List<Product>>())
-            .onErrorResumeNext(new BuyClientExceptionHandler<List<Product>>())
+            .compose(new UnwrapRetrofitBodyTransformer<ProductTagsWrapper, List<ProductTag>>())
+            .map(new Func1<List<ProductTag>, List<String>>() {
+                @Override
+                public List<String> call(final List<ProductTag> productTags) {
+                    if (productTags == null) {
+                        return null;
+                    } else {
+                        final List<String> tags = new ArrayList<>();
+                        for (ProductTag productTag : productTags) {
+                            if (!TextUtils.isEmpty(productTag.getHandle())) {
+                                tags.add(productTag.getHandle());
+                            }
+                        }
+                        return tags;
+                    }
+                }
+            })
+            .onErrorResumeNext(new BuyClientExceptionHandler<List<String>>())
             .observeOn(callbackScheduler);
     }
 
     @Override
-    public CancellableTask getCollections(final int page, final Callback<List<Collection>> callback) {
-        return new CancellableTaskSubscriptionWrapper(getCollections(page).subscribe(new InternalCallbackSubscriber<>(callback)));
+    public CancellableTask getProducts(final int page, final Long collectionId, Set<String> tags, final Collection.SortOrder sortOrder, final Callback<List<Product>> callback) {
+        return new CancellableTaskSubscriptionWrapper(getProducts(page, collectionId, tags, sortOrder).subscribe(new InternalCallbackSubscriber<>(callback)));
     }
 
     @Override
-    public Observable<List<Collection>> getCollections(final int page) {
+    public Observable<List<Product>> getProducts(final int page, final Long collectionId, final Set<String> tags, final Collection.SortOrder sortOrder) {
         if (page < 1) {
             throw new IllegalArgumentException("page is a 1-based index, value cannot be less than 1");
         }
 
-        // All collection responses from the server are wrapped in a CollectionListings object which contains and array of collections
-        // For this call, we will clamp the size of the collection array returned to the page size
+        final String tagsQueryStr = tags != null && !tags.isEmpty() ? TextUtils.join(",", tags.toArray()) : null;
+
+        final String sortOrderStr;
+        if (collectionId != null) {
```

Comment:
Actually, can you make `sortOrder` default to `Collection.SortOrder.COLLECTION_DEFAULT` (or `if (sortOrder == null) sortOrder = DEFAULT`), and initialize `sortOrderStr` to `null` (then won't need the `else` block). Then `if (collectionId != null)` is the only thing we need to check.